src/adaptors/curve/index.js

```diff
@@ -1,269 +1,277 @@
+  const blockchainToPoolSubgraphPromise = Object.fromEntries(
+    BLOCKCHAINIDS.map((blockchainId) => [
+      blockchainId,
+      getSubGraphData(blockchainId),
+    ])
+  );
+  const blockchainToPoolPromise = Object.fromEntries(
+    BLOCKCHAINIDS.filter((blockchainId) => blockchainId !== 'ethereum').map(
+      (blockchainId) => [blockchainId, getPools(blockchainId)]
+    )
   );
 
-  return dataTvl;
-};
+  // we need the ethereum data first for the crv prive and await extra query to CG
+  const ethereumPools = await getPools('ethereum');
+  const priceCrv = getPriceCrv(Object.values(ethereumPools));
+
+  // create feeder closure to fill defillamaPooldata asynchroniously
+  const defillamaPooldata = [];
+  const feedLlama = (poolData, blockchainId) => {
+    const [
+      addressToPool,
+      addressToPoolSubgraph,
+      addressToGauge,
+      gaugeAddressToExtraRewards,
+    ] = poolData;
+    for (const [address, pool] of Object.entries(addressToPool)) {
+      const subgraph = addressToPoolSubgraph[address];
+      const gauge = addressToGauge[blockchainId][pool.gaugeAddress];
+      // one gauge can have multiple (different) extra rewards
+      const extraRewards = gaugeAddressToExtraRewards[pool.gaugeAddress];
+
+      const apyBase = subgraph ? parseFloat(subgraph.latestDailyApy) : 0;
+      const aprCrv =
+        gauge && subgraph ? getPoolAPR(pool, subgraph, gauge, priceCrv) : 0;
+      const aprExtra = extraRewards
+        ? extraRewards.map((reward) => reward.apy).reduce((a, b) => a + b)
+        : 0;
+
+      // tokens are listed using their contract addresses
+      // https://github.com/DefiLlama/yield-server#adaptors
+      const underlyingTokens = pool.coins.map((coin) => coin.address);
+      const rewardTokens = extraRewards
+        ? extraRewards.map((reward) => reward.tokenAddress)
+        : [];
+      if (aprCrv) {
+        rewardTokens.push('0xD533a949740bb3306d119CC777fa900bA034cd52'); // CRV
+      }
 
-const buildPool = (entry, chainString) => {
-  const token = pools.tokenMapping[chainString][entry.pool];
-  const id = pools.pools[chainString].find(
-    (el) => el.name === entry.pool
-  )?.swap;
-  let apy = entry.apy < 0 ? 0 : entry.apy;
-  apy = chainString === 'ethereum' ? apy : apy * 100;
+      const tvlUsd = pool.usdTotal;
+      if (!tvlUsd) {
+        continue;
+      }
 
-  const newObj = {
-    // ugly, but some swap pools have same address on diff networks
-    pool: `${id}-${chainString}`,
-    chain: utils.formatChain(chainString),
-    project: 'curve',
-    symbol: utils.formatSymbol(token),
-    tvlUsd: entry.tvl,
-    apy: apy,
+      defillamaPooldata.push({
+        pool: address + '-' + blockchainId,
+        chain: utils.formatChain(blockchainId),
+        project: 'curve',
+        symbol: pool.coins.map((coin) => coin.symbol).join('-'),
+        tvlUsd,
+        apyBase,
+        apyReward: aprCrv + aprExtra,
+        apy: apyBase + aprCrv + aprExtra,
```

Comment:
not wrong but we could omit the sum as we calculate it on the server before saving to the db

Reply:
the test would need to be adapted. It checks for `apy` _not_ either or as in the docs. I was not sure whether it's needed.

Reply:
hmm, should run, i just tested locally without `apy` field and worked without issues